docs/slightfile-design.md

```diff
@@ -0,0 +1,266 @@
+
+<!-- vscode-markdown-toc -->
+* [Table of Contents](#TableofContents)
+* [Background](#Background)
+* [Problem Statement](#ProblemStatement)
+* [Proposal](#Proposal)
+* [Alternatives](#Alternatives)
+	* [Alternative 1](#Alternative1)
+	* [Alternative 2](#Alternative2)
+	* [Alternative 3](#Alternative3)
+	* [Alternative 4](#Alternative4)
+	* [Alternative 5](#Alternative5)
+* [Additional Details](#AdditionalDetails)
+	* [Edge Cases](#EdgeCases)
+		* [Duplicate Name](#DuplicateName)
+
+<!-- vscode-markdown-toc-config
+	numbering=false
+	autoSave=true
+	/vscode-markdown-toc-config -->
+<!-- /vscode-markdown-toc -->
+
+## <a name='Background'></a>Background
+
+Currently, a common `slightfile` might look like this:
+```toml
+specversion = "0.1"
+secret_store = "configs.azapp"
+
+[[capability]]
+name = "kv.azblob"
+
+[[capability]]
+name = "mq.azsbus"
+```
+
+These capabilities can then be acessed in a user application, like so:
+```rs
+use anyhow::Result;
+
+use kv::*;
+wit_bindgen_rust::import!("../../wit/kv.wit");
+wit_error_rs::impl_error!(kv::Error);
+
+use mq::*;
+wit_bindgen_rust::import!("../../wit/mq.wit");
+wit_error_rs::impl_error!(mq::Error);
+
+fn main() -> Result<()> {
+    let kv = Kv::open("the-name-of-your-container")?;
+    let mq = Mq::open("the-name-of-your-queue")?;
+
+    // -snip-
+}
+```
+
+## <a name='ProblemStatement'></a>Problem Statement
+
+The current design has two main restrictions:
+(1) we do not allow for multiple-capabilities of the same namespace but different implementor types (e.g., `kv.azblob`, and `kv.filesystem`), and
+(2) we do not allow for different capabilities to use different secret stores (i.e., we only have one global secret store).
+
+## <a name='Proposal'></a>Proposal
+
+The changes involve:
+- re-naming the `name` field to `resource`,
+- re-utilization of the `name` field to create a named resource,
+- eliminating the global `secret_store` in favour of being more declarative with configuration variables, and
+- creating a map field (i.e., `configs`) for a dynamic number of configs.
+
+```toml
+specversion = "0.1"
+
+[[capability]]
+resource = "kv.filesystem"
+name = "orders"
+
+[[capability]]
+resource = "kv.azblob"
+name = "customers"
+    [capability.configs]
+    AZURE_STORAGE_ACCOUNT = "{envvars.AZURE_STORAGE_ACCOUNT}"
+    AZURE_STORAGE_KEY = "{envvars.AZURE_STORAGE_KEY}"
+```
+
+The example `slightfile` above can be used in code, like so:
+```rs
+use anyhow::Result;
+
+use kv::*;
+wit_bindgen_rust::import!("../../wit/kv.wit");
+wit_error_rs::impl_error!(kv::Error);
+
+fn main() -> Result<()> {
+    let orders = Kv::open("orders", "orders-folder")?; 
+    let customers = Kv::open("customers", "customers-container")?;
+}
+```
+
+It is important to note that there will be a new field added onto every `open` function relating to the name of that capability. Plus, for capabilities that export mulitple resources, one single name will identify them all. For example:
```

Comment:
I think we may want to separate out container / table portion of the redesign until we discuss this further. Also, I don't think it really affects the `slightfile` changes. 

Perhaps, just use the code example as it currently is without the extra bit about folder / container.

Reply:
Currently, the `open` function for `kv` accepts a `name` that is used differently according to the implementor we are using~ for `kv.filesystem`, `open` requires the name of a folder that will be create to store that store's key-value pairs; on the other hand, for `kv.azblob`, `open` requires the name of a container to access. That's how things currently are.

The example just illustrates that the `open` function will now require another parameter — the name of the resource. That is, the use of that is now:
```rs
  // -snip-
  let my_store = Kv::open("new-param-name-of-the-resource", "existing-param-whatever-name-the-implementor-requires");
  // -snip-
```

I hope that clarifies things.

Reply:
Thank you for the clarification. I confused an earlier conversation.